File: experiments/foreground_exchange/android/app/src/main/java/org/trailmesh/foregroundprobe/ForegroundFrame.kt
package org.trailmesh.foregroundprobe

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.MessageDigest
import java.util.Locale

internal object ForegroundFrame {
    const val MAX_PAYLOAD_BYTES = 16 * 1024
    private const val FRAME_HEADER_BYTES = 4 + 32
    private val dataMagic = "TMD1".toByteArray(Charsets.US_ASCII)
    private val ackMagic = "TMA1".toByteArray(Charsets.US_ASCII)

    data class DecodedFrame(val payload: ByteArray, val digest: ByteArray)
    data class DataMessage(val attemptIndex: Int, val frame: DecodedFrame)
    data class AckMessage(val attemptIndex: Int, val accepted: Boolean, val digest: ByteArray)

    fun makeTestPayload(size: Int, attemptIndex: Int): ByteArray {
        require(size in 1..MAX_PAYLOAD_BYTES) { "payload size is outside the 16 KiB limit" }
        require(attemptIndex >= 0) { "attempt index must be non-negative" }
        return ByteArray(size) { index -> ((index + attemptIndex) % 251).toByte() }
    }

    fun sha256(payload: ByteArray): ByteArray = MessageDigest.getInstance("SHA-256").digest(payload)

    fun sha256Hex(payload: ByteArray): String = hex(sha256(payload))

    fun hex(bytes: ByteArray): String = bytes.joinToString("") {
        "%02x".format(Locale.ROOT, it.toInt() and 0xff)
    }

    fun encode(payload: ByteArray): ByteArray {
        require(payload.size <= MAX_PAYLOAD_BYTES) { "payload exceeds the 16 KiB limit" }
        val buffer = ByteBuffer.allocate(FRAME_HEADER_BYTES + payload.size).order(ByteOrder.BIG_ENDIAN)
        buffer.putInt(payload.size)
        buffer.put(sha256(payload))
        buffer.put(payload)
        return buffer.array()
    }

    fun decode(frame: ByteArray): DecodedFrame {
        require(frame.size >= FRAME_HEADER_BYTES) { "frame is truncated" }
        val buffer = ByteBuffer.wrap(frame).order(ByteOrder.BIG_ENDIAN)
        val size = buffer.int
        require(size in 0..MAX_PAYLOAD_BYTES) { "payload exceeds the 16 KiB limit" }
        require(frame.size == FRAME_HEADER_BYTES + size) { "frame length does not match payload length" }
        val expectedDigest = ByteArray(32).also { buffer.get(it) }
        val payload = ByteArray(size).also { buffer.get(it) }
        val actualDigest = sha256(payload)
        require(MessageDigest.isEqual(expectedDigest, actualDigest)) { "payload checksum mismatch" }
        return DecodedFrame(payload, actualDigest)
    }

    fun encodeDataMessage(attemptIndex: Int, payload: ByteArray): ByteArray {
        require(attemptIndex >= 0) { "attempt index must be non-negative" }
        return ByteBuffer.allocate(dataMagic.size + 4 + FRAME_HEADER_BYTES + payload.size)
            .order(ByteOrder.BIG_ENDIAN)
            .put(dataMagic)
            .putInt(attemptIndex)
            .put(encode(payload))
            .array()
    }

    fun decodeDataMessage(message: ByteArray): DataMessage {
        require(message.size >= dataMagic.size + 4 + FRAME_HEADER_BYTES) { "data message is truncated" }
        val buffer = ByteBuffer.wrap(message).order(ByteOrder.BIG_ENDIAN)
        require(ByteArray(dataMagic.size).also { buffer.get(it) }.contentEquals(dataMagic)) { "unknown message type" }
        val attemptIndex = buffer.int
        require(attemptIndex >= 0) { "attempt index must be non-negative" }
        val frame = ByteArray(buffer.remaining()).also(buffer::get)
        return DataMessage(attemptIndex, decode(frame))
    }

    fun encodeAckMessage(attemptIndex: Int, accepted: Boolean, digest: ByteArray): ByteArray {
        require(attemptIndex >= 0) { "attempt index must be non-negative" }
        require(digest.size == 32) { "SHA-256 digest must be 32 bytes" }
        return ByteBuffer.allocate(4 + 4 + 1 + 32).order(ByteOrder.BIG_ENDIAN)
            .put(ackMagic)
            .putInt(attemptIndex)
            .put(if (accepted) 1 else 0)
            .put(digest)
            .array()
    }

    fun decodeAckMessage(message: ByteArray): AckMessage {
        require(message.size == 41) { "acknowledgement has an invalid size" }
        val buffer = ByteBuffer.wrap(message).order(ByteOrder.BIG_ENDIAN)
        require(ByteArray(4).also(buffer::get).contentEquals(ackMagic)) { "unknown message type" }
        val attemptIndex = buffer.int
        require(attemptIndex >= 0) { "attempt index must be non-negative" }
        val accepted = when (buffer.get().toInt() and 0xff) {
            0 -> false
            1 -> true
            else -> throw IllegalArgumentException("acknowledgement status is invalid")
        }
        return AckMessage(attemptIndex, accepted, ByteArray(32).also { buffer.get(it) })
    }

    fun isAckMessage(message: ByteArray): Boolean = message.size >= ackMagic.size &&
        message.copyOfRange(0, ackMagic.size).contentEquals(ackMagic)
}

File: experiments/foreground_exchange/ios/Sources/ForegroundFrame.swift
import CryptoKit
import Foundation

enum ForegroundFrameError: Error, Equatable {
    case invalidSize
    case invalidAttempt
    case truncated
    case invalidLength
    case checksumMismatch
    case unknownMessage
    case invalidAcknowledgement
}

struct DecodedFrame {
    let payload: Data
    let digest: Data
}

struct ProbeDataMessage {
    let attemptIndex: Int
    let frame: DecodedFrame
}

struct ProbeAckMessage {
    let attemptIndex: Int
    let accepted: Bool
    let digest: Data
}

enum ForegroundFrame {
    static let maxPayloadBytes = 16 * 1024
    private static let frameHeaderBytes = 4 + 32
    private static let dataMagic = Data("TMD1".utf8)
    private static let ackMagic = Data("TMA1".utf8)

    static func makeTestPayload(size: Int, attemptIndex: Int) throws -> Data {
        guard (1...maxPayloadBytes).contains(size) else { throw ForegroundFrameError.invalidSize }
        guard attemptIndex >= 0 else { throw ForegroundFrameError.invalidAttempt }
        return Data((0..<size).map { UInt8(($0 + attemptIndex) % 251) })
    }

    static func digest(_ payload: Data) -> Data {
        Data(SHA256.hash(data: payload))
    }

    static func digestHex(_ payload: Data) -> String {
        hex(digest(payload))
    }

    static func hex(_ bytes: Data) -> String {
        bytes.map { String(format: "%02x", $0) }.joined()
    }

    static func encode(_ payload: Data) throws -> Data {
        guard payload.count <= maxPayloadBytes else { throw ForegroundFrameError.invalidSize }
        var frame = Data()
        appendUInt32(UInt32(payload.count), to: &frame)
        frame.append(digest(payload))
        frame.append(payload)
        return frame
    }

    static func decode(_ frame: Data) throws -> DecodedFrame {
        guard frame.count >= frameHeaderBytes else { throw ForegroundFrameError.truncated }
        let payloadSize = Int(try readUInt32(frame, at: 0))
        guard payloadSize <= maxPayloadBytes,
              frame.count == frameHeaderBytes + payloadSize else {
            throw ForegroundFrameError.invalidLength
        }
        let expectedDigest = Data(frame[4..<36])
        let payload = Data(frame[36..<frame.count])
        let actualDigest = digest(payload)
        guard expectedDigest == actualDigest else { throw ForegroundFrameError.checksumMismatch }
        return DecodedFrame(payload: payload, digest: actualDigest)
    }

    static func encodeDataMessage(attemptIndex: Int, payload: Data) throws -> Data {
        guard attemptIndex >= 0 else { throw ForegroundFrameError.invalidAttempt }
        var message = dataMagic
        appendUInt32(UInt32(attemptIndex), to: &message)
        message.append(try encode(payload))
        return message
    }

    static func decodeDataMessage(_ message: Data) throws -> ProbeDataMessage {
        guard message.count >= 8 + frameHeaderBytes,
              Data(message.prefix(4)) == dataMagic else {
            throw ForegroundFrameError.unknownMessage
        }
        let attemptIndex = Int(try readUInt32(message, at: 4))
        let frame = Data(message.dropFirst(8))
        return ProbeDataMessage(attemptIndex: attemptIndex, frame: try decode(frame))
    }

    static func encodeAckMessage(attemptIndex: Int, accepted: Bool, digest: Data) throws -> Data {
        guard attemptIndex >= 0 else { throw ForegroundFrameError.invalidAttempt }
        guard digest.count == 32 else { throw ForegroundFrameError.invalidAcknowledgement }
        var message = ackMagic
        appendUInt32(UInt32(attemptIndex), to: &message)
        message.append(accepted ? 1 : 0)
        message.append(digest)
        return message
    }

    static func decodeAckMessage(_ message: Data) throws -> ProbeAckMessage {
        guard message.count == 41, Data(message.prefix(4)) == ackMagic else {
            throw ForegroundFrameError.invalidAcknowledgement
        }
        let attemptIndex = Int(try readUInt32(message, at: 4))
        let acceptedByte = message[8]
        guard acceptedByte == 0 || acceptedByte == 1 else {
            throw ForegroundFrameError.invalidAcknowledgement
        }
        return ProbeAckMessage(
            attemptIndex: attemptIndex,
            accepted: acceptedByte == 1,
            digest: Data(message[9..<41]))
    }

    static func isAckMessage(_ message: Data) -> Bool {
        message.count >= 4 && Data(message.prefix(4)) == ackMagic
    }

    private static func appendUInt32(_ value: UInt32, to data: inout Data) {
        var bigEndian = value.bigEndian
        withUnsafeBytes(of: &bigEndian) { data.append(contentsOf: $0) }
    }

    private static func readUInt32(_ data: Data, at offset: Int) throws -> UInt32 {
        guard offset >= 0, data.count >= offset + 4 else { throw ForegroundFrameError.truncated }
        return data[offset..<(offset + 4)].reduce(UInt32(0)) { ($0 << 8) | UInt32($1) }
    }
}
